File: packages/password-input/src/password-input.ts
export interface PasswordInputProps {
  /** Whether the password is currently revealed (shown as plain text). */
  revealed?: boolean
  /**
   * Accessible label for the reveal (show password) action.
   * @default 'Show password'
   */
  revealLabel?: string
  /**
   * Accessible label for the hide action.
   * @default 'Hide password'
   */
  hideLabel?: string
}

export interface PasswordInputAPI {
  /** The `type` attribute for the underlying input given the revealed state. */
  inputType: 'text' | 'password'
  /** The accessible label for the toggle button given the revealed state. */
  toggleLabel: string
  /** The `aria-pressed` value for the toggle button. */
  togglePressed: boolean
}

export const DEFAULT_REVEAL_LABEL = 'Show password'
export const DEFAULT_HIDE_LABEL = 'Hide password'

/**
 * State-free password-input helper. Given a `revealed` flag (owned by the
 * framework adapter), it returns the input `type`, the toggle button's
 * accessible label, and its `aria-pressed` value. It does not hold state.
 */
export function createPasswordInput(props: PasswordInputProps = {}): PasswordInputAPI {
  const {
    revealed = false,
    revealLabel = DEFAULT_REVEAL_LABEL,
    hideLabel = DEFAULT_HIDE_LABEL,
  } = props

  return {
    inputType: revealed ? 'text' : 'password',
    toggleLabel: revealed ? hideLabel : revealLabel,
    togglePressed: revealed,
  }
}

File: packages/flutter/lib/src/components/password_input.dart
import 'package:flutter/material.dart';
import '../theme/refraction_theme.dart';
import 'input.dart';

/// A password field with a built-in reveal/hide toggle.
///
/// Wraps [RefractionInput], toggling the masking of the underlying field
/// between obscured and plain text via a trailing eye button. Mirrors the
/// React/Astro `PasswordInput` adapter — it forwards [validationState],
/// [leadingIcon], [controller], [onChanged] and friends straight through to
/// the underlying input.
///
/// ```dart
/// RefractionPasswordField(
///   placeholder: 'Password',
///   leadingIcon: const Icon(Icons.lock_outline),
///   onChanged: (value) => password = value,
/// )
/// ```
class RefractionPasswordField extends StatefulWidget {
  /// Optional externally-managed controller forwarded to [RefractionInput].
  final TextEditingController? controller;

  /// Optional externally-managed focus node forwarded to [RefractionInput].
  final FocusNode? focusNode;

  /// Hint text shown when the field is empty.
  final String? placeholder;

  /// When true, the field is dimmed and read-only.
  final bool disabled;

  /// Optional leading icon (e.g. a lock) forwarded to [RefractionInput].
  final Widget? leadingIcon;

  /// Optional validation affordance forwarded to [RefractionInput].
  final RefractionInputValidationState? validationState;

  /// Called on every keystroke with the current text.
  final ValueChanged<String>? onChanged;

  /// Called when the user submits the field.
  final ValueChanged<String>? onSubmitted;

  /// Accessible label for the reveal action. Defaults to `'Show password'`.
  final String revealLabel;

  /// Accessible label for the hide action. Defaults to `'Hide password'`.
  final String hideLabel;

  /// Creates a [RefractionPasswordField].
  const RefractionPasswordField({
    super.key,
    this.controller,
    this.focusNode,
    this.placeholder,
    this.disabled = false,
    this.leadingIcon,
    this.validationState,
    this.onChanged,
    this.onSubmitted,
    this.revealLabel = 'Show password',
    this.hideLabel = 'Hide password',
  });

  @override
  State<RefractionPasswordField> createState() =>
      _RefractionPasswordFieldState();
}

class _RefractionPasswordFieldState extends State<RefractionPasswordField> {
  bool _revealed = false;

  void _toggle() => setState(() => _revealed = !_revealed);

  @override
  Widget build(BuildContext context) {
    final colors = RefractionTheme.of(context).colors;

    return RefractionInput(
      controller: widget.controller,
      focusNode: widget.focusNode,
      placeholder: widget.placeholder,
      disabled: widget.disabled,
      leadingIcon: widget.leadingIcon,
      validationState: widget.validationState,
      obscureText: !_revealed,
      onChanged: widget.onChanged,
      onSubmitted: widget.onSubmitted,
      suffix: Semantics(
        button: true,
        label: _revealed ? widget.hideLabel : widget.revealLabel,
        child: MouseRegion(
          cursor: widget.disabled
              ? SystemMouseCursors.forbidden
              : SystemMouseCursors.click,
          child: GestureDetector(
            onTap: widget.disabled ? null : _toggle,
            child: Icon(
              _revealed ? Icons.visibility_off_outlined : Icons.visibility_outlined,
              size: 16,
              color: colors.mutedForeground,
            ),
          ),
        ),
      ),
    );
  }
}
